readthedocs/projects/models.py

```diff
@@ -1465,14 +1477,109 @@ def __str__(self):
         return self.email
 
 
+class WebHookEvent(models.Model):
+
+    BUILD_TRIGGERED = 'build:triggered'
+    BUILD_PASSED = 'build:passed'
+    BUILD_FAILED = 'build:failed'
+
+    EVENTS = (
+        (BUILD_TRIGGERED, _('Build triggered')),
+        (BUILD_PASSED, _('Build passed')),
+        (BUILD_FAILED, _('Build failed')),
+    )
+
+    name = models.CharField(
+        max_length=256,
+        unique=True,
+        choices=EVENTS,
+    )
+
+    def __str__(self):
+        return self.name
+
+
 class WebHook(Notification):
+
     url = models.URLField(
+        _('URL'),
         max_length=600,
         help_text=_('URL to send the webhook to'),
     )
+    secret = models.CharField(
+        help_text=_('Secret used to sign the payload of the webhook'),
+        max_length=255,
+        blank=True,
+        null=True,
+    )
+    events = models.ManyToManyField(
+        WebHookEvent,
+        related_name='webhooks',
+        help_text=_('Events to subscribe'),
+    )
+    payload = models.TextField(
+        _('JSON payload'),
+        help_text=_(
+            'JSON payload to send to the webhook. '
+            'Check the docs for available substitutions.',
+        ),
+        blank=True,
+        null=True,
+        max_length=25000,
+    )
+    exchanges = GenericRelation(
+        'integrations.HttpExchange',
+        related_query_name='webhook',
+    )
+
+    def save(self, *args, **kwargs):
+        if not self.secret:
+            self.secret = get_random_string(length=32)
+        super().save(*args, **kwargs)
+
+    def get_payload(self, version, build, event):
+        """Get the final payload replacing all placeholders."""
+        if not self.payload:
+            return None
+
+        project = version.project
+        organization = project.organizations.first()
+
+        organization_name = ''
+        organization_slug = ''
+        if organization:
+            organization_slug = organization.slug
+            organization_name = organization.name
+
+        substitutions = {
+            '${event}': event,
+            '${build.id}': build.id,
+            '${build.commit}': build.commit or '',
```

Comment:
was thinking we could have `${build.commit.short}` to display only the 7/8 first chars of the commit.